### ISSUE TITLE
Include a filter function with a rejection list by default

### ISSUE THREAD

**Issue by shakesoda** [bakpakin/tiny-ecs#2]
It's pretty common (in my experience) to need a system to process all entities that match a given set of components _unless_ it's got specific other components. i.e. I don't want my raycasting system to bother processing whatever the player is possessing, or it'll result in always hitting itself. I think it'd be nice to include something like this, so I wouldn't have to copy/paste/rewrite this function every time I start a new project.

Here's the function I've been using

``` lua
-- Usage: system.filter = require_all_unless({"accept", "these"}, {"unless", "it", "has_these"})
local function require_all_unless(components, reject)
    local len = #components
    local rlen = #reject
    assert(type(components) == "table")
    assert(type(reject) == "table")
    return function(_, e)
        local c
        for i = 1, len do
            c = components[i]
            if type(c) == 'function' then
                if not c(_, e) then
                    return false
                end
            elseif e[c] == nil then
                return false
            end
        end
        -- same thing, but reversed.
        for i = 1, rlen do
            c = reject[i]
            if type(c) == 'function' then
                if c(_, e) then
                    return false
                end
            elseif e[c] ~= nil then
                return false
            end
        end
        return true
    end
end
```

**Comment by s-ol** [comment]
To make this a little more general: I would build a method `filter.all_except(..., reject)`, having a vararg of required components and `reject` being either: 
- a string 
- a numerically indexed table of strings
- another filter

That would be the most flexible and intuitive API in my opinion.

**Comment by shakesoda** [comment]
@S0lll0s the ... has to be last, though, doesn't it? I thought (reject, ...) would be a bit weird, so I made it two tables.

**Comment by s-ol** [comment]
I think it has to be syntactically but you can define it as (...) and `select` the last argument.

```
function all_except(...)
  local all = {...}
  local except = table.remove(all) -- pop last argument
  if type(except) == "string" then
    except = {except}
  end
  if type(except) == "table" then
    except = filter.all(unpack(except))
  end
  all = filter.all(unpack(all))
  return function(e)
   return all(e) and not except(e)
  end
end
```

**Comment by bakpakin** [comment]
Good idea, that should be in the API.

As for what the syntax should be, I have an idea. It's not extensively marked in the documentation, but Filters generated by `tiny.requireAll` and `tiny.requireOne` can be nested. For example, 

``` lua
local myFilter = tiny.requireAll(tiny.requireOne("This", "orThis"), tiny.requireOne("AndThis", "orThisOne"))
```

This is flexible and should probably be made more explicit in the docs. With a function like `tiny.reject` or `tiny.exclude`, you could easily nest these into more complicated Filters.

**Comment by s-ol** [comment]
I had that in mind, but wanted to expand on that above (by allowing filters/functions and strings as arguments). I think wherever ambiguity can be avoided and no or next-to-no overhead is involved a more flexible API is better (Ruby approach basically) but I don't know if that fits your personal design for the API.

**Comment by shakesoda** [comment]
@S0lll0s ah, simple workaround.
@bakpakin that _definitely_ needs to be documented better! I like that solution. Maybe rejectOne and rejectAll to be consistent with the requires?

btw, I see you're using {...} all over the place for varargs, it is worth noting that I recently did some benchmarks on that versus select("#", ...) + select(i, ...) and the selects are _very_ significantly faster on LuaJIT due to not running into any NYI stuff.

**Comment by bakpakin** [comment]
How about a general filter function that takes up to three arguments -- a list of required Components, a list of one-required Components, and a list of excluded Components? 

Maybe:

``` lua
tiny.filter(
  {"requiresThis", "andThis"}, 
  {"excludesThis", "andThisOne"}, 
  {"RequireOneOfThese", "a", "b", "c"}
)
```

If any of these parameters were left as nil, the function would infer an empty list. This would provide very similar syntax to shakesoda's original suggestion.

**Comment by bakpakin** [comment]
@shakesoda  As for using `select` over `{...}`, that seems simple enough and I see no reason not to.

I'm definitely going to add the reject filters - I like `rejectOne` and `rejectAll`, and maybe add the general filter.

I want Filters to be flexible, so I'm willing to add more filter functions if need be.

**Comment by shakesoda** [comment]
Sounds good to me! :+1:

**Comment by bakpakin** [comment]
Released 1.1-2, with two new filter functions, rejectAll and rejectAny. Also renamed requireOne to requireAny.
